spec/index.bs

```diff
@@ -4249,103 +4248,92 @@ The full binding address of a resource can be defined as a trio of:
 </div>
 
 Issue: should this example and the note be moved to some "best practices" document?
 
-Note: the expected usage of the {{GPUPipelineLayout}} is placing the most common and the least frequently changing bind groups at the "bottom" of the layout, meaning lower bind group slot numbers, like 0 or 1. The more frequently a bind group needs to change between draw calls, the higher its index should be. This general guideline allows the user agent to minimize state changes between draw calls, and consequently lower the CPU overhead.
+Note: the expected usage of the [=pipeline layout=] is placing the most common and the least frequently changing bind groups at the "bottom" of the layout, meaning lower bind group slot numbers, like 0 or 1. The more frequently a bind group needs to change between draw calls, the higher its index should be. This general guideline allows the user agent to minimize state changes between draw calls, and consequently lower the CPU overhead.
+
+<div algorithm class=validusage>
+    A [=pipeline layout=] |layout| is
+    <dfn abstract-op lt="validate pipeline layout for limits">valid for the limits of</dfn> {{GPUDevice}} |device|
+    if and only if the following requirements are met:
+
+    1. Let |allEntries| be the result of concatenating
+        |bgl|.{{GPUBindGroupLayout/[[descriptor]]}}.{{GPUBindGroupLayoutDescriptor/entries}}
+        for all |bgl| in |layout|.{{pipeline layout/[[bindGroupLayouts]]}}.
+    1. |allEntries| must not [=exceeds the binding slot limits|exceed the binding slot limits=] of
+        |device|.{{GPUObjectBase/[[device]]}}.{{device/[[limits]]}}.
+</div>
 
 ### Creation ### {#pipeline-layout-creation}
 
-A {{GPUPipelineLayout}} is created via {{GPUDevice/createPipelineLayout()|GPUDevice.createPipelineLayout()}}.
+[=Pipeline layouts=] are created by the [$computed pipeline layout$] algorithm
+(inside {{GPUDevice/createComputePipeline()}} and {{GPUDevice/createRenderPipeline()}}),
+to match the specified <dfn dictionary>GPUPipelineLayoutRequirements</dfn>:
 
 <script type=idl>
-dictionary GPUPipelineLayoutDescriptor : GPUObjectDescriptorBase {
-    required sequence<GPUBindGroupLayout> bindGroupLayouts;
+dictionary GPUPipelineLayoutRequirements {
+    required sequence<(GPUBindGroupLayout or GPUAutoBindGroupLayoutMode)> bindGroupLayouts;
 };
-</script>
 
-<dl dfn-type=method dfn-for=GPUDevice>
-    : <dfn>createPipelineLayout(descriptor)</dfn>
-    ::
-        Creates a {{GPUPipelineLayout}}.
-
-        <div algorithm=GPUDevice.createPipelineLayout>
-            **Called on:** {{GPUDevice}} |this|.
-
-            **Arguments:**
-            <pre class=argumentdef for="GPUDevice/createPipelineLayout(descriptor)">
-                |descriptor|: Description of the {{GPUPipelineLayout}} to create.
-            </pre>
-
-            **Returns:** {{GPUPipelineLayout}}
+enum GPUAutoBindGroupLayoutMode {
+    "auto",
```

Comment:
Does using `auto` for a group that has zero bindings in the shader create an empty BGL, or does it creates no BGL? This is currently different, see https://github.com/gpuweb/gpuweb/issues/2043.

Reply:
As implemented I think you get exactly one BGL per entry in bindGroupLayouts, so in this case you get an empty BGL. I think this is reasonable behavior.

I guess getBindGroupLayout needs to deal with this as well, it still assumes the bindGroupLayouts is filled to the max with empty BGLs.

Reply:
It could return to you an empty BGL in that case as well, anyway that's a detail that will need fleshing out after/if this PR lands.